pkg/queue/postgres/retention.go

```diff
@@ -0,0 +1,164 @@
+)
+
+// retentionTaskSpec defines a SQL task to remove completed tasks that match given criteria.
+type retentionTaskSpec struct {
+	handlers.SQLExecTaskSpec
+	QueueName string           `json:"queueName"`
+	TaskType  queue.TaskType   `json:"taskType"`
+	Status    queue.TaskStatus `json:"status"`
+	Age       time.Duration    `json:"age"`
+}
+
+// NewRetentionHandler creates a task handler that will clean up old finished tasks
+func NewRetentionHandler(db *sql.DB) queue.TaskHandler {
+	return handlers.NewSQLTaskHandler("RetentionHandler", db)
+}
+
+// AssertRetentionSchedule creates a new queue retention tasks for the supplied queue, finished tasks matching
+// the supplied parameters will be deleted
+func AssertRetentionSchedule(ctx context.Context, db *sql.DB, queueName string, taskType queue.TaskType, status queue.TaskStatus, age time.Duration) (err error) {
+	span, ctx := opentracing.StartSpanFromContext(ctx, "AssertRetentionSchedule")
+	span.SetTag("pkg.name", "postgres")
+
+	spec := createRetentionSpec(queueName, taskType, status, age)
+	specBytes, err := json.Marshal(spec)
+	if err != nil {
+		return fmt.Errorf("can not build retention task spec: %w", err)
+	}
+	// randomly distribute the retention tasks throughout the hour
+	when := rand.Intn(60)
+	retentionSchedule := queue.TaskScheduleRequest{
+		TaskBase: queue.TaskBase{
+			Queue: MaintenanceTaskQueue,
+			Type:  RetentionTask,
+			Spec:  specBytes,
+		},
+		CronSchedule: fmt.Sprintf("%d * * * *", when), // every hour at minute "when"
+	}
+	tx, err := db.BeginTx(ctx, nil)
+	if err != nil {
+		return fmt.Errorf("can not start transaction for scheduling: %w", err)
+	}
+	defer func() {
+		if err != nil {
+			_ = tx.Rollback()
+			return
+		}
+		err = tx.Commit()
+	}()
+
+	_, err = tx.ExecContext(ctx, `LOCK TABLE schedules IN ACCESS EXCLUSIVE MODE;`)
+	if err != nil {
+		return fmt.Errorf("failed to lock `schedules`: %w", err)
+	}
+
+	builder := squirrel.StatementBuilder.
+		PlaceholderFormat(squirrel.Dollar).
+		RunWith(cdb.WrapWithTracing(tx))
+
+	var exists int
+	// use a unique error name here otherwise the sql.ErrNoRows might shadow
+	// us and things will break. This is also handled by the named error return
+	// variable, but this makes the code easier to copy and paste
+	existsErr := builder.Select("1").
+		From("schedules").
+		Where(squirrel.Eq{
+			"task_queue":              MaintenanceTaskQueue,
+			"task_type":               RetentionTask,
+			"task_spec->>'queueName'": queueName,
+			"task_spec->>'taskType'":  taskType,
+			"task_spec->>'status'":    status,
+		}).ScanContext(ctx, &exists)
+	if existsErr != nil && existsErr != sql.ErrNoRows {
+		return fmt.Errorf("can not verify existing schedule: %w", existsErr)
+	}
+
+	// will only non-zero if err is nil and task is not found
+	if exists == 0 {
+		span.SetTag("created", true)
+		// pass nil db because it doesn't need the raw db
+		return NewScheduler(nil).Schedule(ctx, builder, retentionSchedule)
```

Comment:
Perhaps we should remove the database from the function completely and feed the builder into all the scheduler functions, this will break the interface though. Since we're already breaking it here I think it makes sense to do this in the same release.

Reply:
🤔 the only method that is using the db is the `AssertTask`. https://github.com/contiamo/go-base/blob/7a4d4a3a0e3992eb7cf3c3b269cf1730be697b16/pkg/queue/scheduler.go#L28

It occurs to me that passing the builder into the method exposes that it _must_ be a sql based queue. We could not replace it easily with another queue without changing all of the interfaces.

I also wonder if it is safe to call this lock multiple times in the same transaction 
https://github.com/contiamo/go-base/blob/7a4d4a3a0e3992eb7cf3c3b269cf1730be697b16/pkg/queue/postgres/scheduler.go#L179

Reply:
Seems that calling the same lock within the same transaction is fine 
```sql
hub=# BEGIN;
BEGIN
hub=# LOCK TABLE schedules IN ACCESS EXCLUSIVE MODE;
LOCK TABLE
hub=# SELECT * FROM schedules;
hub=# LOCK TABLE schedules IN ACCESS EXCLUSIVE MODE;
LOCK TABLE
hub=# LOCK TABLE schedules IN ACCESS EXCLUSIVE MODE;
LOCK TABLE
hub=# LOCK TABLE schedules IN ACCESS EXCLUSIVE MODE;
LOCK TABLE
hub=# COMMIT;
```

Reply:
I'm not sure about safety either.